# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/base/Pivot.cpp
# Commit: c6db418af190
# Full Hash: c6db418af19097f562d94b87bbf25362556b2fbc
# Author: Morgan Rae Reschenberg <mreschenberg@mozilla.com>
# Date: 2022-10-21 08:39:18
# Regressor Bug: 1787284
# File Overlap Count: 1
# Description:
#   Bug 1787284: [Part 2] Cache name and create radio name pivot rule r=eeejay
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D159118
# ==============================================================================

diff -r fda6706bb580 -r c6db418af190 accessible/base/Pivot.cpp
--- a/accessible/base/Pivot.cpp	Thu Oct 20 21:40:05 2022 +0000
+++ b/accessible/base/Pivot.cpp	Thu Oct 20 21:40:06 2022 +0000
@@ -634,3 +634,29 @@
   }
   return nsIAccessibleTraversalRule::FILTER_MATCH;
 }
+
+// Radio Button Name Rule
+
+PivotRadioNameRule::PivotRadioNameRule(const nsString& aName) : mName(aName) {}
+
+uint16_t PivotRadioNameRule::Match(Accessible* aAcc) {
+  uint16_t result = nsIAccessibleTraversalRule::FILTER_IGNORE;
+  RemoteAccessible* remote = aAcc->AsRemote();
+  if (!remote || !StaticPrefs::accessibility_cache_enabled_AtStartup()) {
+    // We need the cache to be able to fetch the name attribute below.
+    return result;
+  }
+
+  if (nsAccUtils::MustPrune(aAcc)) {
+    result |= nsIAccessibleTraversalRule::FILTER_IGNORE_SUBTREE;
+  }
+
+  if (remote->IsHTMLRadioButton()) {
+    nsString currName = remote->GetCachedHTMLRadioNameAttribute();
+    if (!currName.IsEmpty() && mName.Equals(currName)) {
+      result |= nsIAccessibleTraversalRule::FILTER_MATCH;
+    }
+  }
+
+  return result;
+}